### PR TITLE
[pipelining] Lower _configure_data_parallel_mode to stage

### PR DESCRIPTION
**Pull request by kwen2501**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #127946
* #127935

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @d4l3k

### CODE REVIEW

File: torch/distributed/pipelining/PipelineStage.py [hunk 1]
```diff
@@ -464,8 +453,17 @@ def forward_maybe_with_nosync(self, *args, **kwargs):
         return out_val
 
     def backward_maybe_with_nosync(self, bwd_kwargs: Dict, bwd_chunk_id: int):
+        """
+        Whether using PP with FSDP or DDP, there are some runtime differences between the last backward step and the
+        other steps.  Namely, we need to accumulate gradients on previous steps and reduce them on the last step, but
+        there are additional state-variables and performance considerations depending on the data parallelism used.
+        This helper should adapt any pipeline parallel schedule to work with common/supported data parallel libraries.
+        """
+        last_backward = bwd_chunk_id == self.chunks - 1
```
Comment (wconstab) [inline torch/distributed/pipelining/PipelineStage.py:462]: isn't this broken if the schedule tries to do the bwd chunks in the reverse order as @H-Huang was proposing for bfs?

File: torch/distributed/pipelining/PipelineStage.py [hunk 2]
```diff
@@ -442,17 +442,6 @@ def _retrieve_recv_grads(
         grads = self._map_tensor_from_recv_info(recv_infos)
         return grads
 
-    def _configure_data_parallel_mode(self, last_backward: bool):
```
[the remark this answers is not in the document]
Comment (kwen2501) [reply torch/distributed/pipelining/PipelineStage.py:445]: True.
In reality, the `if isinstance(self.submod, FSDPModule):` line needs to be lifted to the parent caller, so not two much modularity left to make this a function.

File: torch/distributed/pipelining/PipelineStage.py (hunk 1 above)
Comment (kwen2501) [reply torch/distributed/pipelining/PipelineStage.py:462]: Discussed and not an issue now.

Comment (wconstab) [reply torch/distributed/pipelining/PipelineStage.py:462]: uh.. it is still an issue! in the previous code, the schedule was controlling 'data parallel mode' so it was the schedule's responsibility to make sure that it got called correctly.

unless we actively assert that schedules always call bwd_chunks in linear order (0..N), which we don't, then this adds a little risk exposure.  Should we at least assert this?

Comment (kwen2501) [reply torch/distributed/pipelining/PipelineStage.py:462]: You caught me. Maybe we still need a `seen_bwd_chunks` counter in `PipelineStage`. And:
```
last_backward = seen_bwd_chunks == self.chunks - 1
```

Comment (H-Huang) [reply torch/distributed/pipelining/PipelineStage.py:462]: > last_backward = seen_bwd_chunks == self.chunks - 1

If we go this route then we wouldn't need to pass bwd_chunk_id into `backward_maybe_with_nosync` right?

File: torch/distributed/pipelining/PipelineStage.py [hunk 3]
```diff
@@ -447,9 +442,18 @@ def forward_maybe_with_nosync(self, *args, **kwargs):
...
+        This helper should adapt any pipeline parallel schedule to work with common/supported data parallel libraries.
+        """
+        last_backward = self._seen_bwd_chunks == self.chunks - 1
+
+        # If submod is wrapped by DDP
         if isinstance(self.submod, DistributedDataParallel):
-            if bwd_chunk_id == self.chunks - 1:
+            if last_backward:
                 # Last chunk, prepare for gradient reduction
                 # HACK: reaching into DDP implementation details here. Is there a better way?
```
Comment (wconstab) [inline torch/distributed/pipelining/PipelineStage.py:458]: nit: remove HACK comment? we are doing the same thing for FSDP, there probably isn't a better way, or if there is we should say what it is :)

**Review by wconstab — APPROVED** [review]
lgtm

**Pull request closed without merging**